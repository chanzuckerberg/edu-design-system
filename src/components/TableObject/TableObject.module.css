@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    #TABLE OBJECT
\*------------------------------------*/

/**
  * 1) A table object is a container for a data table
  *    and toolbars either above or below the table
  * 2) Important for uses like filtering and pagination
  */
.table-object {
  width: 100%;
  overflow-x: hidden;
  border-radius: var(--eds-theme-border-radius);
}

/**
  * Table object body
  */
.table-object__body {
  position: relative;

  /**
    * Table object body with content overflowed left
    */
  &.eds-is-overflow-left {
    /**
     * Right overflow gradient for the overflow list
     */
    /* stylelint-disable-next-line max-nesting-depth */
    &:before {
      content: '';
      display: block;
      position: absolute;
      background: linear-gradient(
        to left,
        rgba(255, 255, 255, 0.001) 50%,
        rgba(0, 0, 0, 0.1)
      );
      height: 100%;
      top: 0;
      left: 0;
      width: var(--eds-size-4);
      z-index: var(--eds-z-index-100);
    }
  }

  /**
    * Table object body with content overflowed right
    */
  &.eds-is-overflow-right {
    /**
     * Right overflow gradient for the overflow list
     */
    /* stylelint-disable-next-line max-nesting-depth */
    &:after {
      content: '';
      display: block;
      position: absolute;
      background: linear-gradient(
        to right,
        rgba(255, 255, 255, 0.001) 50%,
        rgba(0, 0, 0, 0.1)
      );
      height: 100%;
      top: 0;
      right: 0;
      width: var(--eds-size-4);
      z-index: var(--eds-z-index-100);
    }
  }
}

/**
  * Table object body inner
  * 1) Overflow auto to allow scrolling when too much content
  */
.table-object__body-inner {
  width: 100%;
  overflow: auto; /* 1 */
}
