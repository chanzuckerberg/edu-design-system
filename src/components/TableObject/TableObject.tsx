import clsx from 'clsx';
import React, { ReactNode } from 'react';
import styles from './TableObject.module.css';

export interface Props {
  /**
   * Child node(s) that can be nested inside component
   */
  children: ReactNode;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
}

/**
 * Primary UI component for user interaction
 */
export const TableObject: React.FC<Props> = ({
  children,
  className,
  ...other
}) => {
  const componentClassName = clsx(styles['table-object'], className, {});
  return (
    <div className={componentClassName} {...other}>
      {children}
    </div>
  );
};
