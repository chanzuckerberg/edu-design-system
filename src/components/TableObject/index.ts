export {TableObject as default} from './TableObject';
