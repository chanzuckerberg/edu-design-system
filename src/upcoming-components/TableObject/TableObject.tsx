import clsx from 'clsx';
import React, { ReactNode } from 'react';
import styles from './TableObject.module.css';
import TableObjectBody from '../TableObjectBody';
import TableObjectHeader from '../TableObjectHeader';

export interface Props {
  /**
   * Child node(s) that can be nested inside component
   */
  children: ReactNode;
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
}

/**
 * Primary UI component for user interaction
 */
export const TableObject = ({ children, className, ...other }: Props) => {
  const componentClassName = clsx(styles['table-object'], className, {});
  return (
    <div className={componentClassName} {...other}>
      {children}
    </div>
  );
};

TableObject.Body = TableObjectBody;
TableObject.Header = TableObjectHeader;
