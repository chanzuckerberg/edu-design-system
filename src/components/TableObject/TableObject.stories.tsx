import { Story, Meta } from '@storybook/react';
import React from 'react';

import { TableObject, Props } from './TableObject';
import Table from '../Table';
import TableBody from '../TableBody';
import TableCell from '../TableCell';
import TableHeader from '../TableHeader';
import TableObjectBody from '../TableObjectBody';
import TableObjectHeader from '../TableObjectHeader';
import TableRow from '../TableRow';
import Toolbar from '../Toolbar';
import ToolbarItem from '../ToolbarItem';

export default {
  title: 'Organisms/Tables/Table Object',
  component: TableObject,
} as Meta;

const Template: Story<Props> = (args) => (
  <TableObject>
    <TableObjectHeader>
      <Toolbar>
        <ToolbarItem>
          <div className="fpo">Toolbar Item</div>
        </ToolbarItem>
      </Toolbar>{' '}
    </TableObjectHeader>
    <TableObjectBody>
      <Table caption="This is a table caption and it is required">
        <TableHeader>
          <TableRow>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableObjectBody>
  </TableObject>
);

const OverflowLargeTemplate: Story<Props> = (args) => (
  <TableObject>
    <TableObjectHeader>
      <Toolbar>
        <ToolbarItem>
          <div className="fpo">Toolbar Item</div>
        </ToolbarItem>
      </Toolbar>{' '}
    </TableObjectHeader>
    <TableObjectBody>
      <Table
        caption="This is a table caption and it is required"
        behavior="overflow-lg"
      >
        <TableHeader>
          <TableRow>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableObjectBody>
  </TableObject>
);

const OverflowSmallTemplate: Story<Props> = (args) => (
  <TableObject>
    <TableObjectHeader>
      <Toolbar>
        <ToolbarItem>
          <div className="fpo">Toolbar Item</div>
        </ToolbarItem>
      </Toolbar>{' '}
    </TableObjectHeader>
    <TableObjectBody>
      <Table
        caption="This is a table caption and it is required"
        behavior="overflow-sm"
      >
        <TableHeader>
          <TableRow>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableObjectBody>
  </TableObject>
);

export const Default = Template.bind({});
Default.args = {};

export const OverflowLarge = OverflowLargeTemplate.bind({});
OverflowLarge.args = {};

export const OverflowSmall = OverflowSmallTemplate.bind({});
OverflowSmall.args = {};
