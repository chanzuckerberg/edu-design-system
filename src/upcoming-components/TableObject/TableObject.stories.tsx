import { BADGE } from '@geometricpanda/storybook-addon-badges';
import { Story, Meta } from '@storybook/react';
import React from 'react';

import { TableObject, Props } from './TableObject';
import Table from '../../components/Table';
import TableBody from '../../components/TableBody';
import TableCell from '../../components/TableCell';
import TableHeader from '../../components/TableHeader';
import TableRow from '../../components/TableRow';
import Toolbar from '../../components/Toolbar';
import ToolbarItem from '../../components/ToolbarItem';
import TableObjectBody from '../TableObjectBody';
import TableObjectHeader from '../TableObjectHeader';

export default {
  title: 'Organisms/Tables/Table Object',
  component: TableObject,
  subcomponents: { TableObjectBody, TableObjectHeader },
  parameters: {
    axe: {
      // TODO: re-enable when component is worked on
      skip: true,
    },
    badges: [BADGE.BETA],
  },
} as Meta;

const Template: Story<Props> = (args) => (
  <TableObject>
    <TableObject.Header>
      <Toolbar>
        <ToolbarItem>
          <div className="fpo">Toolbar Item</div>
        </ToolbarItem>
      </Toolbar>{' '}
    </TableObject.Header>
    <TableObject.Body>
      <Table caption="This is a table caption and it is required">
        <TableHeader>
          <TableRow>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableObject.Body>
  </TableObject>
);

const OverflowTemplate: Story<Props> = (args) => (
  <TableObject>
    <TableObject.Header>
      <Toolbar>
        <ToolbarItem>
          <div className="fpo">Toolbar Item</div>
        </ToolbarItem>
      </Toolbar>{' '}
    </TableObject.Header>
    <TableObject.Body>
      <Table
        behavior="overflow"
        caption="This is a table caption and it is required"
      >
        <TableHeader>
          <TableRow>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
            <TableCell as="th">Table heading</TableCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>

            <TableCell>Value</TableCell>
            <TableCell>Value</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableObject.Body>
  </TableObject>
);

export const Default = Template.bind({});
Default.args = {};

export const Overflow = OverflowTemplate.bind({});
Overflow.args = {};
