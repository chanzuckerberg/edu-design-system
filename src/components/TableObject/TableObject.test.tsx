import {generateSnapshots} from '@chanzuckerberg/story-utils';
import * as stories from './TableObject.stories';

describe('<TableObject />', () => {
  generateSnapshots(stories);
});
