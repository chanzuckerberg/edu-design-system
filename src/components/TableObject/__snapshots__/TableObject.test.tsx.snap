// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableObject /> Default story renders snapshot 1`] = `
<div
  class="table-object"
>
  <div
    class="table-object__header"
  >
    <div
      class="toolbar"
    >
      <div
        class="toolbar__item"
      >
        <h2
          class="heading heading--size-h3"
        >
          Checkpoint Progress
        </h2>
      </div>
      <div
        class="toolbar__item toolbar__item--align-right"
      >
        <button
          class="clickable-style clickable-style--md clickable-style--secondary clickable-style--brand button button--secondary"
          data-bootstrap-override="clickable-style-secondary"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            height="1.375rem"
            style="--icon-size: 1.375rem;"
            width="1.375rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="test-file-stub#filter-list"
            />
          </svg>
          Filters
        </button>
      </div>
    </div>
  </div>
  <div
    class="table-object__body"
  >
    <div
      class="table-object__body-inner"
    >
      <table
        class="table"
      >
        <caption
          class="table__caption u-is-vishidden"
        >
          This is a table caption and it is required
        </caption>
        <thead
          class="table__header"
        >
          <tr
            class="table__row"
          >
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body"
        >
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`<TableObject /> Overflow story renders snapshot 1`] = `
<div
  class="table-object"
>
  <div
    class="table-object__header"
  >
    <div
      class="toolbar"
    >
      <div
        class="toolbar__item"
      >
        <h2
          class="heading heading--size-h3"
        >
          Checkpoint Progress
        </h2>
      </div>
      <div
        class="toolbar__item toolbar__item--align-right"
      >
        <button
          class="clickable-style clickable-style--md clickable-style--secondary clickable-style--brand button button--secondary"
          data-bootstrap-override="clickable-style-secondary"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            height="1.375rem"
            style="--icon-size: 1.375rem;"
            width="1.375rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="test-file-stub#filter-list"
            />
          </svg>
          Filters
        </button>
      </div>
    </div>
  </div>
  <div
    class="table-object__body"
  >
    <div
      class="table-object__body-inner"
    >
      <table
        class="table"
      >
        <caption
          class="table__caption u-is-vishidden"
        >
          This is a table caption and it is required
        </caption>
        <thead
          class="table__header"
        >
          <tr
            class="table__row"
          >
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Table heading
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body"
        >
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`<TableObject /> Sortable story renders snapshot 1`] = `
<div
  class="table-object"
>
  <div
    class="table-object__header"
  >
    <div
      class="toolbar"
    >
      <div
        class="toolbar__item"
      >
        <h2
          class="heading heading--size-h3"
        >
          Checkpoint Progress
        </h2>
      </div>
      <div
        class="toolbar__item toolbar__item--align-right"
      >
        <button
          class="clickable-style clickable-style--md clickable-style--secondary clickable-style--brand button button--secondary"
          data-bootstrap-override="clickable-style-secondary"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="icon"
            fill="currentColor"
            height="1.375rem"
            style="--icon-size: 1.375rem;"
            width="1.375rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <use
              xlink:href="test-file-stub#filter-list"
            />
          </svg>
          Filters
        </button>
      </div>
    </div>
  </div>
  <div
    class="table-object__body"
  >
    <div
      class="table-object__body-inner"
    >
      <table
        class="table"
      >
        <caption
          class="table__caption u-is-vishidden"
        >
          This is a table caption and it is required
        </caption>
        <thead
          class="table__header"
        >
          <tr
            class="table__row"
          >
            <th
              class="table__header-cell"
            >
              <button
                class="table__header-cell-button"
              >
                Sortable
                <div
                  class="table__header-cell-sort"
                >
                  <svg
                    class="icon"
                    fill="currentColor"
                    height="1rem"
                    role="img"
                    style="--icon-size: 1rem;"
                    width="1rem"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <title
                      id="4"
                    >
                      Sort
                    </title>
                    <use
                      xlink:href="test-file-stub#unfold-more"
                    />
                  </svg>
                </div>
              </button>
            </th>
            <th
              class="table__header-cell"
            >
              <div
                class="table__header-cell-contents"
              >
                Not sortable
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body"
        >
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
          <tr
            class="table__row"
          >
            <td
              class="table__cell"
            >
              Value
            </td>
            <td
              class="table__cell"
            >
              Value
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
